modules/io/src/writers/xviz-protobuf-writer.js

```diff
@@ -0,0 +1,248 @@
+  }
+
+  _writeMessageIndex() {
+    this._checkValid();
+    const {startTime, endTime, messages} = this.messageTimings;
+    const messageTimings = {};
+
+    if (startTime) {
+      messageTimings.startTime = startTime;
+    }
+
+    if (endTime) {
+      messageTimings.endTime = endTime;
+    }
+
+    // Sort messages by index before writing out as an array
+    const messageTimes = Array.from(messages.keys()).sort((a, b) => a - b);
+
+    const timing = [];
+    messageTimes.forEach((value, index) => {
+      // Value is two greater than message index
+      const limit = timing.length;
+      if (value > limit) {
+        // Adding 2 because 1-frame is metadata file, so message data starts at 2
+        throw new Error(
+          `Error writing time index file. Messages are missing between ${limit + 2} and ${value +
+            2}`
+        );
+      }
+
+      timing.push(messages.get(value));
+    });
+    messageTimings.timing = timing;
+
+    const msg = JSON.stringify(messageTimings);
+    this.writeToSink('0-frame.json', msg);
+    this.wroteMessageIndex = timing.length;
+  }
+
+  close() {
+    if (this.sink) {
+      if (!this.wroteMessageIndex) {
+        this._writeMessageIndex();
+      }
+
+      super.close();
+    }
+  }
+
+  /* eslint-disable camelcase */
+  _saveTimestamp(xviz_data, index) {
+    const {log_info, updates} = xviz_data;
+
+    if (index === undefined) {
+      // Metadata case
+      if (log_info) {
+        const {start_time, end_time} = log_info || {};
+        if (start_time) {
+          this.messageTimings.startTime = start_time;
+        }
+
+        if (end_time) {
+          this.messageTimings.endTime = end_time;
+        }
+      }
+    } else if (updates) {
+      if (updates.length === 0 || !updates.every(update => typeof update.timestamp === 'number')) {
+        throw new Error('XVIZ updates did not contain a valid timestamp');
+      }
+
+      const min = Math.min(updates.map(update => update.timestamp));
+      const max = Math.max(updates.map(update => update.timestamp));
+
+      this.messageTimings.messages.set(index, [min, max, index, messageName(index)]);
+    } else {
+      // Missing updates & index is invalid call
+      throw new Error('Cannot find timestamp');
+    }
+  }
+  /* eslint-enable camelcase */
+
+  writeToSink(name, msg) {
+    this.sink.writeSync(name, msg);
+  }
+}
+
+const COLOR_KEYS = ['stroke_color', 'fill_color'];
+/* Convert color to a flattened array */
+function toColorArray(object) {
+  const clrs = object.substring(1);
+  const len = clrs.length;
+  if (!(len === 3 || len === 4 || len === 6 || len === 8)) {
+    return null;
+  }
+
+  const color = [];
+  const step = clrs.length === 3 || clrs.length === 4 ? 1 : 2;
+  for (let i = 0; i < clrs.length; i += step) {
+    color.push(parseInt(clrs.substr(i, step), 16));
+  }
+
+  return color;
+}
+
+// Recursively walk object performing the following conversions
```

Comment:
Maybe explain what this is converting the protobuf **into**, I think it's the JSON format right?

Reply:
Actually, its the JSON format from builder works for other encodings but Protobuf is more strict (doesn't not support variants such as fields with both string and array values in JSON).  So this normalize for protobuf encoding, which is what I'll explain in the comment.